compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/ConeTypeCompatibilityChecker.kt

```diff
@@ -0,0 +1,402 @@
+    private data class BoundTypeArgument(val type: ConeKotlinType, val variance: Variance)
+
+    /** Accumulated type arguments bound to a type parameter declared in a class. */
+    private data class BoundTypeArguments(val upper: MutableSet<ConeClassLikeType>, val lower: MutableSet<ConeClassLikeType>)
+
+    private fun ConeClassLikeType.toConcreteClass(ctx: ConeInferenceContext): ConcreteClass? {
+        return lookupTag.toConcreteClass(ctx)
+    }
+
+    private fun ConeClassLikeLookupTag.toConcreteClass(
+        ctx: ConeInferenceContext
+    ): ConcreteClass? = when (val klass = ctx.symbolProvider.getSymbolByLookupTag(this)?.fir) {
+        is FirTypeAlias -> klass.getAliasedClass(ctx)?.let { ConcreteClass.TypeAlias(klass, it, ctx) }
+        is FirClass<*> -> ConcreteClass.Class(klass, ctx)
+        else -> null
+    }
+
+    private sealed class ConcreteClass {
+        data class Class(val firClass: FirClass<*>, override val ctx: ConeInferenceContext) : ConcreteClass()
+        data class TypeAlias(val firTypeAlias: FirTypeAlias, val referencedClass: FirClass<*>, override val ctx: ConeInferenceContext) :
+            ConcreteClass()
+
+        abstract val ctx: ConeInferenceContext
+
+        val isInterface: Boolean
+            get() {
+                return when (this) {
+                    is Class -> firClass.isInterface
+                    is TypeAlias -> referencedClass.isInterface
+                }
+            }
+
+        val superConcreteClasses: Set<ConcreteClass> by lazy {
+            val firClass: FirClass<*> = when (this) {
+                is Class -> firClass
+                is TypeAlias -> referencedClass
+            }
+            firClass.superConeTypes.mapNotNull { it.lookupTag.toConcreteClass(ctx) }.toSet()
+        }
+
+        @OptIn(ExperimentalStdlibApi::class)
+        val thisAndAllSuperConcreteClasses: Set<ConcreteClass> by lazy {
+            val queue = ArrayDeque<ConcreteClass>()
+            queue.addLast(this)
+            buildSet {
+                add(this@ConcreteClass)
+                while (queue.isNotEmpty()) {
+                    val current = queue.removeFirst()
+                    val superTypes = current.superConcreteClasses
+                    queue.addAll(superTypes - this@buildSet)
+                    addAll(superTypes)
+                }
+            }
+        }
+
+        val isFinal: Boolean
+            get() {
+                return when (this) {
+                    is Class -> firClass.isFinal()
+                    is TypeAlias -> referencedClass.isFinal()
+                }
+            }
+
+        val isEnum: Boolean
+            get() {
+                val firClass = when (this) {
+                    is Class -> firClass
+                    is TypeAlias -> referencedClass
+                }
+                return firClass.isEnumClass || firClass.classId == StandardClassIds.Enum
+            }
+
+        private fun FirClass<*>.isFinal(): Boolean {
```

Comment:
Thanks for all of those `val`-ifications. :) I wonder, though, this can be `val` too.